File: apps/web/src/hoodchart.ts
// A year panel drawn as a picture instead of a table (Kyle, 2026-09-22: "for the longitudinal data, we should give
// the user the option on page to switch between a table and chart view"). Same numbers, same file, same rules.
//
// **Lines on one chart, since later the same day.** Kyle, after looking at the first bars: he wants "Homes sold"
// and "Building permits" on the SAME chart, with the two selectable. So: one pair of axes, one line per series
// that is switched on, a key of real checkboxes, and every year on the axis.
//
// **Every number is the real number, since later still the same day** (Kyle: "I want exact numbers"; DECISIONS
// 2026-09-22). There is no hidden count anywhere in the dataset any more, so there is no marker for one: a year
// has a value or has nothing recorded, and that is the whole model.
//
// The rules docs/13 sets, and how each one survives being drawn:
//
//  1. **No ranking, no comparison with another neighborhood.** A chart holds ONE neighborhood's own years. There
//     is no city line, no trend line, no second place on the axis, and no colour that means good or bad: the
//     colours are identities (homes, permits; walking, biking, badly hurt), not two ends of a scale. One y-axis,
//     never two — two scales on one chart is a way of making any two lines say whatever you like, so a series in
//     a different UNIT (days to close) gets a chart of its own, never a second axis.
//  2. **A year with nothing recorded is a break in the line, never a zero.** Joining across one would draw a
//     number nobody counted.
//  3. **The table is the source of truth.** The chart is another VIEW of the same rows, never a different set of
//     numbers, and the table stays in the page (visually hidden) so a screen reader has every value either way
//     (apps/web/src/hoods.ts, `yearGroup`).
//  4. **Three ways to tell lines apart, at once.** Colour, point shape (circle, diamond, square) and line pattern
//     (solid, dashed, dash-dot), so up to three series are still three in grayscale, in print, under forced
//     colours and to a reader who sees no difference between the hues. Never more than three on one chart.
//
// This file is arithmetic and strings. It is the model the iPhone (`HoodChart` in HelpCore/HoodChart.swift) and
// Android (`hoodChartModel` in Hoods.kt) re-implement against the same cases, so all three draw the same picture.

/** One year of one series, exactly as the bundle gives it: a number, or nothing recorded. */
export interface ChartPoint { year: string; count: number | undefined; partial: boolean }

/** Which colour, which marker shape, which dash. Identities, never a scale. At most three on one chart. */
export type Tone = 'a' | 'b' | 'c';
/** What the numbers are: a count of things, or a number of days. A chart holds ONE unit (rule 1). */
export type Unit = 'count' | 'days';

/** One series before it is measured against the others. `key` is what a checkbox switches. */
export interface ChartSeries {
  key: string;
  tone: Tone;
  /** The series' own name — "Homes sold", "Permits" — which is its key entry and its checkbox label. */
  label: string;
  points: ChartPoint[];
  /** `count` unless said otherwise. */
  unit?: Unit;
}

/** A year of a series, placed. `frac` is its height as a share of the top of the axis, 0 to 1. */
export interface PlotPoint {
  year: string;
  index: number;
  partial: boolean;
  kind: 'value' | 'none';
  /** The number, when there is one to show. */
  value?: number;
  /** 0 to 1. */
  frac: number;
}

/** A piece of the line between two neighbouring years. */
export interface PlotSegment { from: number; to: number }

export interface SeriesModel {
  key: string;
  tone: Tone;
  label: string;
  points: PlotPoint[];
  segments: PlotSegment[];
  /** The years with nothing recorded at all: a place on the axis, no marker, and a break in the line. */
  blanks: string[];
  /** The tallest year, for the summary sentence. `null` when nothing can be drawn. */
  peak: { year: string; value: number } | null;
}

export interface ChartModel {
  years: string[];
  series: SeriesModel[];
  /** The top of the axis: the first round number at or above the tallest value, never under 1. */
  top: number;
  /** 0, then every labelled value, ending at `top`. */
  ticks: number[];
  unit: Unit;
}

/** 1, 2, 5, 10, 20, 50, … — the only step sizes an axis is allowed, so ticks are always round numbers. */
function axisStep(max: number): number {
  for (let e = 0; e < 12; e++) for (const s of [1, 2, 5]) { const step = s * 10 ** e; if (Math.ceil(max / step) <= 4) return step; }
  return 10 ** 12;
}

/**
 * The whole picture, worked out from the rows — no pixels yet, so the same arithmetic runs in a test, on a phone
 * and in a browser. Years arrive in the order the table lists them and stay in it. Only the series that are
 * switched on are handed in, so the axis follows what is actually on the screen. Every series handed in shares
 * one unit; the first one's unit is the chart's.
 */
export function chartModel(series: readonly ChartSeries[]): ChartModel {
  const years = series[0]?.points.map((p) => p.year) ?? [];
  const max = Math.max(0, ...series.flatMap((s) => s.points.map((p) => (typeof p.count === 'number' ? p.count : 0))));
  const step = axisStep(Math.max(max, 1));
  const top = Math.max(1, Math.ceil(max / step) * step);
  const ticks: number[] = [];
  for (let t = 0; t <= top + 1e-9; t += step) ticks.push(t);
  if (ticks[ticks.length - 1] !== top) ticks.push(top);

  return {
    years, top, ticks, unit: series[0]?.unit ?? 'count',
    series: series.map((s) => {
      const points: PlotPoint[] = s.points.map((p, index) =>
        typeof p.count === 'number'
          ? { year: p.year, index, partial: p.partial, kind: 'value', value: p.count, frac: p.count / top }
          : { year: p.year, index, partial: p.partial, kind: 'none', frac: 0 });
      // The line runs between two neighbouring years whenever both have a value. A year with nothing recorded
      // breaks it, because joining across one would draw a number nobody counted.
      const segments: PlotSegment[] = [];
      for (let i = 0; i + 1 < points.length; i++) if (points[i]!.kind === 'value' && points[i + 1]!.kind === 'value') segments.push({ from: i, to: i + 1 });
      const values = points.filter((p) => p.kind === 'value');
      return {
        key: s.key, tone: s.tone, label: s.label, points, segments,
        blanks: points.filter((p) => p.kind === 'none').map((p) => p.year),
        // The tallest year, and the EARLIEST of them when two are equal, so one bundle always names one year.
        peak: values.reduce<{ year: string; value: number } | null>(
          (best, p) => (best && best.value >= p.value! ? best : { year: p.year, value: p.value! }), null),
      };
    }),
  };
}

/**
 * Whether a series is worth offering a chart of at all: three years with a number in them. Two points is a line
 * between two dots, not a shape worth a control.
 */
export function chartable(points: readonly ChartPoint[]): boolean {
  return points.filter((p) => typeof p.count === 'number').length >= 3;
}

/** Whether a series has anything at all: one year with a number. Otherwise the panel says the City has not
 *  published it for this neighborhood, instead of a table of "none recorded". */
export function anyValue(points: readonly ChartPoint[]): boolean {
  return points.some((p) => typeof p.count === 'number');
}

/**
 * Which series are drawn, given what has been switched off. **Never nothing**: the last one left on cannot be
 * switched off, so the chart is never an empty pair of axes. The control for it is disabled as well, and the key
 * says why (`hood.chart_only_one`).
 */
export function shownSeries<T extends { key: string }>(all: readonly T[], off: ReadonlySet<string>): T[] {
  const on = all.filter((s) => !off.has(s.key));
  return on.length ? on : [all[0]!];
}

/**
 * Which years get a label under the axis: every one while they fit, then every other one once there are more than
 * six, so a year is never drawn over its neighbour at a large font scale.
 *
 * It counts back from the LAST year rather than forward from the first, because the last year is the one a reader
 * looks for and because stepping forward and then adding the last as well puts two labels side by side whenever
 * the count is even — "2025 2026" ran into each other at a font scale of 2 (Android emulator, 2026-09-22). The
 * first year is added back only when it is not next to the earliest label kept. Every year is still in the table,
 * in each point's own label, and named in the summary sentence, which always says the whole range.
 */
export function axisYears(years: readonly string[]): string[] {
  if (years.length <= 6) return [...years];
  const keep = new Set<number>();
  for (let i = years.length - 1; i >= 0; i -= 2) keep.add(i);
  if (!keep.has(0) && Math.min(...keep) >= 2) keep.add(0);
  return years.filter((_, i) => keep.has(i));
}

// ---- drawing ----------------------------------------------------------------------------------------------------
// Inline SVG and nothing else: no canvas (a canvas carries no `<title>`, nothing for a screen reader, and prints
// as a blur), no library, no script. Everything below is geometry in the viewBox's own units.

const W = 320, GUT = 34, TOP_PAD = 12, BOTTOM = 22, PLOT_H = 96;
const H = TOP_PAD + PLOT_H + BOTTOM;
const MARKER_R = 4;

interface ChartUi {
  t: (key: string, p?: Record<string, string | number>) => string;
  esc: (s: unknown) => string;
}

/**
 * The point at one year. A circle for the first series, a diamond for the second and a square for the third — a
 * different SHAPE, not only a different colour, so the lines are still distinct in grayscale, in print and under
 * forced colours (WCAG 1.4.1).
 */
function markerPath(tone: Tone, x: number, y: number): string {
  return tone === 'a'
    ? `M${x - MARKER_R} ${y}a${MARKER_R} ${MARKER_R} 0 1 0 ${MARKER_R * 2} 0a${MARKER_R} ${MARKER_R} 0 1 0 ${-MARKER_R * 2} 0Z`
    : tone === 'b'
      ? `M${x} ${y - MARKER_R}L${x + MARKER_R} ${y}L${x} ${y + MARKER_R}L${x - MARKER_R} ${y}Z`
      : `M${x - MARKER_R + 0.5} ${y - MARKER_R + 0.5}h${MARKER_R * 2 - 1}v${MARKER_R * 2 - 1}h${-(MARKER_R * 2 - 1)}Z`;
}

/** A value in the chart's unit, as words: "14", or "12 days". */
export function fmtValue(ui: ChartUi, unit: Unit, n: number): string {
  return unit === 'days' ? ui.t('hood.days', { n }) : String(n);
}

/**
 * The whole chart: one pair of axes, one line per series switched on, and a point at every year with a value.
 *
 * Each point is a focusable group with its own `<title>` and its own accessible name, and a small label that
 * appears on hover and on keyboard focus — so "2023, Homes sold: 14" can be reached with a pointer, with a
 * keyboard, and by a screen reader. Every number is also still in the table on the same page.
 */
export function chartSvg(ui: ChartUi, m: ChartModel): string {
  const E = (x: unknown) => ui.esc(x);
  const n = Math.max(1, m.years.length);
  const left = GUT + 8, right = W - 10;
  const xOf = (i: number) => (n === 1 ? (left + right) / 2 : left + (i * (right - left)) / (n - 1));
  const baseline = TOP_PAD + PLOT_H;
  const yOf = (frac: number) => baseline - frac * PLOT_H;
  const labelled = new Set(axisYears(m.years));

  const grid = m.ticks.map((t) => {
    const y = round(yOf(t / m.top));
    return `<line class="gl" x1="${GUT - 4}" y1="${y}" x2="${W - 2}" y2="${y}"></line>` +
      `<text class="tk" x="${GUT - 7}" y="${round(y + 4)}" text-anchor="end">${E(t)}</text>`;
  }).join('');

  const lines = m.series.map((s) => s.segments.map((g) => {
    const a = s.points[g.from]!, b = s.points[g.to]!;
    return `<line class="ln t${s.tone}" x1="${round(xOf(a.index))}" y1="${round(yOf(a.frac))}" x2="${round(xOf(b.index))}" y2="${round(yOf(b.frac))}"></line>`;
  }).join('')).join('');

  const dots = m.series.map((s) => s.points.filter((p) => p.kind === 'value').map((p) => {
    const x = round(xOf(p.index)), y = round(yOf(p.frac)), label = pointText(ui, s.label, p, m.unit);
    const tipX = round(Math.min(Math.max(x, left + 24), right - 24));
    const tip = fmtValue(ui, m.unit, p.value!);
    return `<g class="pt" tabindex="0" role="img" aria-label="${E(label)}"><title>${E(label)}</title>` +
      `<path class="mk t${s.tone}" d="${markerPath(s.tone, x, y)}"></path>` +
      `<g class="tip" aria-hidden="true"><rect x="${round(tipX - 24)}" y="${round(y - 21)}" width="48" height="15" rx="3"></rect>` +
      `<text x="${tipX}" y="${round(y - 10)}" text-anchor="middle">${E(tip)}</text></g></g>`;
  }).join('')).join('');

  const axis = m.years.map((y, i) => labelled.has(y)
    ? `<text class="yr" x="${round(xOf(i))}" y="${baseline + 15}" text-anchor="middle">${E(y)}</text>` : '').join('');

  return `<svg class="hoodsvg" viewBox="0 0 ${W} ${H}" role="group" aria-label="${E(ui.t(m.unit === 'days' ? 'hood.chart_plot_days' : 'hood.chart_plot'))}">
      <g class="grid">${grid}</g>
      <line class="ax" x1="${GUT - 4}" y1="${baseline}" x2="${W - 2}" y2="${baseline}"></line>
      ${lines}${dots}${axis}
    </svg>`;
}

/** What one point says on its own: "2023, Homes sold: 14", "2021, Time to close: 12 days". */
export function pointText(ui: ChartUi, label: string, p: PlotPoint, unit: Unit = 'count'): string {
  const year = p.partial ? ui.t('hood.so_far', { year: p.year }) : p.year;
  const count = p.kind === 'value' ? fmtValue(ui, unit, p.value!) : ui.t('hood.none_recorded');
  return ui.t('hood.chart_bar', { year, label, count });
}

/**
 * The sentence under the picture, in words: what is drawn, over which years, and the biggest year of each series.
 * It is a fact about this neighborhood's own years and never a trend ("rising", "improving"): docs/13's fourth
 * honesty rule is that these numbers describe and do not explain.
 */
export function chartSummary(ui: ChartUi, m: ChartModel): string {
  const parts = m.series.map((s) => s.peak
    ? ui.t('hood.chart_peak', { label: s.label, count: fmtValue(ui, m.unit, s.peak.value), year: s.peak.year })
    : ui.t('hood.chart_peak_none', { label: s.label }));
  return ui.t('hood.chart_summary', {
    from: m.years[0] ?? '', to: m.years[m.years.length - 1] ?? '', most: parts.join(ui.t('list.sep')),
  });
}

/** Two decimals at most: the same geometry from the same numbers on every browser, and a smaller page. */
const round = (x: number) => Math.round(x * 100) / 100;

File: apps/ios/Sources/HelpCore/HoodChart.swift
// A year panel drawn as a picture instead of a table (Kyle, 2026-09-22: "for the longitudinal data, we should give
// the user the option on page to switch between a table and chart view for relevant data").
//
// **Lines on one chart, since later the same day.** Kyle, after looking at the first bars: he wants "Homes sold"
// and "Building permits" on the SAME chart with the two selectable, and a year whose count is hidden to be VISIBLE
// rather than dropped.
//
// This is the Swift half of `apps/web/src/hoodchart.ts`, and, like the rest of HelpCore, it is a port rather than a
// rewrite: the same arithmetic, the same axis, the same wording, held to the same cases in HoodChartTests. It draws
// nothing — no SwiftUI here — so `swift test` runs every rule in it on Linux too.
//
// The rules docs/13 sets, and how each survives being drawn:
//
//  1. **No ranking, no comparison with another neighborhood.** A chart holds ONE neighborhood's own years. There is
//     no whole-city line, no trend line, and no colour that means good or bad: two colours are two identities
//     (homes sold, permits), never two ends of a scale. ONE y-axis, never two.
//  2. **A hidden count is never drawn at a value.** `lt5` — which since 2026-09-22 means blight tickets,
//     demolitions, reported problems and fires, but no longer home sales or building permits (DECISIONS) — becomes
//     a HOLLOW marker at a fixed height, so the year is visibly there and its number is visibly not.
//     `markerFraction` is the only thing that turns a hidden count into geometry. The line goes on through it as a
//     dotted piece, so a hidden year is never a gap that reads as zero.
//  3. **The axis never labels a value under 5 as if it were exact**: `ticks` holds 0 and then nothing below 5, and
//     a marker always sits below the first tick over zero.
//  4. **The table is the source of truth.** The chart is another VIEW of the same rows; the table is still on the
//     screen for VoiceOver either way (HelpApp/HoodsScreen.swift).
import Foundation

/// Which way a neighborhood's year panels are drawn. **Table is the default**, on every platform.
public enum HoodViewChoice: String, Sendable, CaseIterable, Codable {
    case table, chart
}

/// Whatever was stored, read safely: only the exact word "chart" is chart; anything else is a table.
public func hoodViewChoice(_ stored: String?) -> HoodViewChoice { stored == "chart" ? .chart : .table }

public enum HoodChart {

    // MARK: - what goes in

    /// One year of one series, exactly as the bundle gives it: a number, the hidden `lt5`, or nothing recorded.
    public struct Point: Equatable, Sendable, Identifiable {
        public var year: String
        public var count: HoodCount?
        public var partial: Bool
        public var id: String { year }
        public init(year: String, count: HoodCount?, partial: Bool = false) {
            self.year = year; self.count = count; self.partial = partial
        }
    }

    /// Which of the two identities a series wears: a colour, a marker shape and a line pattern, all three.
    public enum Tone: String, Sendable, Equatable { case a, b }

    /// One series before it is measured against the others. `key` is what a checkbox switches.
    public struct Series: Equatable, Sendable, Identifiable {
        public var key: String
        public var tone: Tone
        public var label: String
        public var points: [Point]
        public var id: String { key }
        public init(key: String, tone: Tone, label: String, points: [Point]) {
            self.key = key; self.tone = tone; self.label = label; self.points = points
        }
    }

    // MARK: - what comes out

    public enum Kind: String, Sendable, Equatable { case value, hidden, none }

    /// A year of a series, placed. `frac` is its height as a share of the top of the axis, 0 to 1.
    public struct PlotPoint: Equatable, Sendable, Identifiable {
        public var year: String
        public var index: Int
        public var partial: Bool
        public var kind: Kind
        /// The number, when there is one to show. Never set for a hidden year — that is the point of it.
        public var value: Int?
        /// For a hidden year it is `markerFraction`, a constant chosen by nothing in the data.
        public var frac: Double
        public var id: String { year }
    }

    /// A piece of the line between two neighbouring years. `dotted` where one of its ends is a hidden count.
    public struct Segment: Equatable, Sendable, Identifiable {
        public var from: Int
        public var to: Int
        public var dotted: Bool
        public var id: Int { from }
    }

    public struct SeriesModel: Equatable, Sendable, Identifiable {
        public var key: String
        public var tone: Tone
        public var label: String
        public var points: [PlotPoint]
        public var segments: [Segment]
        /// The years the pipeline hid. A hollow marker each — never a value.
        public var markers: [String]
        /// The years with nothing recorded at all: a place on the axis, no marker, and a break in the line.
        public var blanks: [String]
        /// The tallest year, for the summary sentence; `nil` when there is nothing to draw.
        public var peak: (year: String, value: Int)?
        public var id: String { key }

        public static func == (l: SeriesModel, r: SeriesModel) -> Bool {
            l.key == r.key && l.tone == r.tone && l.label == r.label && l.points == r.points
                && l.segments == r.segments && l.markers == r.markers && l.blanks == r.blanks
                && l.peak?.year == r.peak?.year && l.peak?.value == r.peak?.value
        }
    }

    public struct Model: Equatable, Sendable {
        public var years: [String]
        public var series: [SeriesModel]
        /// The top of the axis. At least 5, so the drawn area never implies a scale finer than suppression.
        public var top: Int
        /// 0, then every labelled value. Never 1 to 4 (rule 3).
        public var ticks: [Int]
    }

    /**
     How high a hidden year's marker sits, as a share of the plot: a CONSTANT — ten of the web's ninety-six drawing
     units — and the same on all three apps, so nothing about the marker comes from the count it stands for. It is
     always below the first tick over zero, because the axis step is at least a quarter of the top.
     */
    public static let markerFraction = 10.0 / 96.0
    public static func markerValue(top: Int) -> Double { Double(top) * markerFraction }

    // MARK: - the model

    /// 1, 2, 5, 10, 20, 50, … — the only step sizes an axis may use, so a tick is always a round number.
    static func axisStep(_ max: Int) -> Int {
        var step = 1
        for _ in 0..<12 {
            for s in [1, 2, 5] {
                let candidate = s * step
                if (max + candidate - 1) / candidate <= 4 { return candidate }
            }
            step *= 10
        }
        return step
    }

    /// The whole picture, worked out from the rows — no geometry yet, so the same arithmetic runs in a test and on
    /// a phone. Only the series that are switched on are handed in, so the axis follows what is on the screen.
    public static func model(_ series: [Series]) -> Model {
        let years = series.first?.points.map(\.year) ?? []
        let max = series.flatMap { $0.points.compactMap { $0.count?.shown } }.max() ?? 0
        let step = axisStep(Swift.max(max, 1))
        let top = Swift.max(5, ((max + step - 1) / step) * step)
        var ticks: [Int] = []
        var t = 0
        while t <= top { if t == 0 || t >= 5 { ticks.append(t) }; t += step }
        if ticks.last != top { ticks.append(top) }

        return Model(years: years, series: series.map { s in
            let points: [PlotPoint] = s.points.enumerated().map { index, p in
                if let n = p.count?.shown {
                    return PlotPoint(year: p.year, index: index, partial: p.partial, kind: .value, value: n,
                                     frac: Double(n) / Double(top))
                }
                if p.count == .suppressed {
                    return PlotPoint(year: p.year, index: index, partial: p.partial, kind: .hidden, value: nil,
                                     frac: markerFraction)
                }
                return PlotPoint(year: p.year, index: index, partial: p.partial, kind: .none, value: nil, frac: 0)
            }
            // The line runs between two neighbouring years whenever both have something. A year with nothing
            // recorded breaks it, because joining across one would draw a number nobody counted.
            var segments: [Segment] = []
            for i in 0..<Swift.max(0, points.count - 1) {
                let a = points[i], b = points[i + 1]
                if a.kind == .none || b.kind == .none { continue }
                segments.append(Segment(from: i, to: i + 1, dotted: a.kind == .hidden || b.kind == .hidden))
            }
            let values = points.filter { $0.kind == .value }
            return SeriesModel(
                key: s.key, tone: s.tone, label: s.label, points: points, segments: segments,
                markers: points.filter { $0.kind == .hidden }.map(\.year),
                blanks: points.filter { $0.kind == .none }.map(\.year),
                // The tallest year, and the EARLIEST of them when two are equal, so one bundle says one year.
                peak: values.reduce(nil) { best, p in
                    (best.map { $0.value >= (p.value ?? 0) } ?? false) ? best : (year: p.year, value: p.value ?? 0)
                }
            )
        }, top: top, ticks: ticks)
    }

    /// Whether a series is worth offering a chart of at all: three years with something in them, and at least one
    /// of those a number there is a point to draw. Two points is a line between two dots.
    public static func chartable(_ points: [Point]) -> Bool {
        points.filter { $0.count != nil }.count >= 3 && points.contains { $0.count?.shown != nil }
    }

    /// Which series are drawn, given what has been switched off. **Never nothing**: the last one left on cannot be
    /// switched off, so the chart is never an empty pair of axes, and its control is disabled with a line saying why.
    public static func shown(_ all: [Series], off: Set<String>) -> [Series] {
        let on = all.filter { !off.contains($0.key) }
        return on.isEmpty ? Array(all.prefix(1)) : on
    }

    /// Which years get a label under the axis: every one while they fit, then every other one once there are more
    /// than six, so a year is never drawn over its neighbour at a large text size.
    ///
    /// It counts back from the LAST year rather than forward from the first: stepping forward and then adding the
    /// last as well puts two labels side by side whenever the count is even, and "2025 2026" ran into each other
    /// at a font scale of 2. The first year is added back only when it is not next to the earliest label kept.
    public static func axisYears(_ years: [String]) -> [String] {
        guard years.count > 6 else { return years }
        var keep = Set<Int>()
        var i = years.count - 1
        while i >= 0 { keep.insert(i); i -= 2 }
        if !keep.contains(0), (keep.min() ?? 0) >= 2 { keep.insert(0) }
        return years.enumerated().filter { keep.contains($0.offset) }.map(\.element)
    }

    /// One count of one neighborhood, year by year, in the order the table prints — the web's `seriesOf`.
    public static func series(_ h: Hood, _ d: Indicators, count: (HoodYear) -> HoodCount?) -> [Point] {
        d.years.map { y in
            Point(year: y, count: count(h.years[y] ?? HoodYear()), partial: Int(y) == d.partialYear)
        }
    }

    // MARK: - the words

    /// What one point says on its own: "2023, Homes sold: 14", "2021, Torn down: fewer than 5". The sentences are
    /// the app's own, handed in, exactly as `HoodFormat.count` takes them.
    public static func pointText(_ p: PlotPoint, label: String, words: (String, [String: String]) -> String) -> String {
        let year = p.partial ? words("hood.so_far", ["year": p.year]) : p.year
        let count: String
        switch p.kind {
        case .value: count = HoodFormat.grouped(Double(p.value ?? 0))
        case .hidden: count = words("hood.lt5", [:])
        case .none: count = words("hood.none_recorded", [:])
        }
        return words("hood.chart_bar", ["year": year, "label": label, "count": count])
    }

    /// The sentence under the picture: what is drawn, over which years, the biggest year of each series, and — only
    /// when there is one — that some years are hidden and are NOT drawn at a value. A fact about this
    /// neighborhood's own years, never a trend: these numbers describe and do not explain (docs/13, rule 4).
    public static func summary(_ m: Model, words: (String, [String: String]) -> String) -> String {
        let most = m.series.map { s -> String in
            guard let peak = s.peak else { return words("hood.chart_peak_none", ["label": s.label]) }
            return words("hood.chart_peak", ["label": s.label,
                                             "count": HoodFormat.grouped(Double(peak.value)),
                                             "year": peak.year])
        }.joined(separator: words("list.sep", [:]))
        let line = words("hood.chart_summary", ["from": m.years.first ?? "", "to": m.years.last ?? "", "most": most])
        guard m.series.contains(where: { !$0.markers.isEmpty }) else { return line }
        return line + " " + words("hood.chart_lt5_note", [:])
    }
}
